web/src/features/datasets/server/service.ts

```diff
@@ -0,0 +1,329 @@
+import { paginationZod, Prisma } from "@langfuse/shared";
+import { prisma } from "@langfuse/shared/src/db";
+import { v4 } from "uuid";
+import { z } from "zod";
+import {
+  clickhouseClient,
+  clickhouseCompliantRandomCharacters,
+  commandClickhouse,
+  convertToScore,
+  type FetchScoresReturnType,
+  logger,
+  queryClickhouse,
+} from "@langfuse/shared/src/server";
+import { aggregateScores } from "@/src/features/scores/lib/aggregateScores";
+import Decimal from "decimal.js";
+
+export const datasetRunsTableSchema = z.object({
+  projectId: z.string(),
+  datasetId: z.string(),
+  queryClickhouse: z.boolean().optional().default(false),
+  ...paginationZod,
+});
+
+type PostgresRunItem = {
+  trace_id: string;
+  observation_id: string;
+  ri_id: string;
+};
+
+type PostgresDatasetRun = {
+  run_id: string;
+  run_name: string;
+  run_description: string;
+  run_metadata: Prisma.JsonValue;
+  run_created_at: Date;
+  run_updated_at: Date;
+  run_items: PostgresRunItem[];
+};
+
+export type DatasetRunsTableInput = z.infer<typeof datasetRunsTableSchema>;
+
+// we might have many traces / observations in Postgres which belong to data in clickhouse.
+// We need to create a temp table in CH, dump the data in there, and then join in CH.
+export const createDatasetRunsTable = async (input: DatasetRunsTableInput) => {
+  const tableName = `dataset_runs_${clickhouseCompliantRandomCharacters()}`;
+  const clickhouseSession = v4();
+  try {
+    const runs = await getDatasetRunsFromPostgres(input);
+
+    await createTempTableInClickhouse(tableName, clickhouseSession);
+    await insertPostgresDatasetRunsIntoClickhouse(
+      runs,
+      tableName,
+      input.projectId,
+      input.datasetId,
+      clickhouseSession,
+    );
+
+    // these calls need to happen sequentially as there can be only one active session with
+    // the same session_id at the time.
+    const scores = await getScoresFromTempTable(
+      input,
+      tableName,
+      clickhouseSession,
+    );
+    const obsAgg = await getObservationLatencyAndCostForDataset(
+      input,
+      tableName,
+      clickhouseSession,
+    );
+    const traceAgg = await getTraceLatencyAndCostForDataset(
+      input,
+      tableName,
+      clickhouseSession,
+    );
+
+    await deleteTempTableInClickhouse(tableName);
+
+    const enrichedRuns = runs.map(({ run_items, ...run }) => {
+      const observation = obsAgg.find((o) => o.runId === run.run_id);
+      const trace = traceAgg.find((t) => t.runId === run.run_id);
```

Comment:
Depending on the cardinality from obsAgg and traceAgg it might be more efficient to create an in-memory map first instead of searching the list each time.